specification/flag-evaluation/flag-evaluation.md

```diff
@@ -0,0 +1,195 @@
+
+```
+// example detailed boolean flag evaluation
+FlagEvaluationDetails<boolean> myBoolDetails = client.getBooleanDetails('bool-flag', false);
+
+// example detailed string flag evaluation
+FlagEvaluationDetails<string> myStringDetails = client.getStringDetails('string-flag', 'N/A', evaluationContext, options);
+
+```
+
+##### Requirement 1.11
+
+> The `evaluation details` wrapper **MUST** have a `value` field matching the type of the flag value, and containing to the flag value.
+
+##### Condition 1.12
+
+> The language supports generics (or an equivalent feature).
+>
+> ##### Conditional Requirement 1.12.1
+>
+> > The `evaluation details` wrapper **SHOULD** accept a generic argument (or utilize an equivalent language feature) which indicates the type of the wrapped `value` field.
+
+##### Requirement 1.13
+
+> The `evaluation details` wrapper **MUST** have a `flag key` field field of type string, representing the flag key.
+
+##### Requirement 1.14
+
+> The `evaluation details` wrapper **MUST** have an optional `variant` field of type string which represents a space-constrained identifier for a particular flag value.
+
+##### Requirement 1.15
+
+> The `evaluation details` wrapper **MUST** have a `reason` field of type string indicating the semantic reason for the returned flag value. Possible values vary by provider, but may include `"TARGETING_MATCH"`, `"SPLIT"`, `"DISABLED"`, `"DEFAULT"`, `"UNKNOWN"` and `"ERROR"`.
+
+##### Requirement 1.16
+
+> In cases of abnormal execution (network failure, unhandled error, etc) the `reason` field in the `evaluation details` **MUST** be set to `"ERROR"`.
+
+##### Requirement 1.17
+
+> The `evaluation details` wrapper **MUST** have an optional `error code` field of type string, which, in cases of abnormal execution, identifies an error occurred during flag evaluation, with possible values `"PROVIDER_NOT_READY"`, `"FLAG_NOT_FOUND"`, `"PARSE_ERROR"`, `"TYPE_MISMATCH"`, or `"GENERAL"`.
+
+##### Requirement 1.18
+
+> If the `reason` in the `evaluation details` is set to `"ERROR"`, the sibling `error code` field **MUST** be set.
+
+##### Requirement 1.19
+
+> The `evaluation details` wrapper **MUST** have a `executed hooks` field, containing an object which itself contains 4 arrays, identified with keys (`before`, `after`, `error` and `finally`) indicating the hooks run for the respective life-cycle stages.
+
+```
+executedHooks: {
+  before: [],
+  after: [],
+  error: [],
+  finally: []
+}
+```
+
+##### Requirement 1.20
+
+> The value of the elements in the arrays contained in the `executed hooks` object **SHOULD** allow authors to uniquely identify the hooks executed during the flag evaluation lifecycle. This could be done with a mandatory string field defined on the `hook` interface, or by using the fully qualified class name in languages where such concepts exist.
```

Comment:
This seems like a feature that I would not include in version 1. If there is need for it then it can always be added later, but for now it seems like trying to cover all the bases before you know if you are even playing baseball.

Reply:
@beeme1mr I guess I agree with Dan's point here.

Reply:
It doesn't have to be in the spec but it was a valuable feature in our experimental project. It allows you to understand what hooks ran and in what order.

Reply:
After some more thought, I agree. It's a good idea but not necessary in the initial spec.